### ISSUE TITLE
Is it possible to train a model using only one 4090?

### ISSUE THREAD

**Issue by yangtianhang** [karpathy/nanochat#157]
I have a 4090 card on my PC. If all the procedure can run using only one 4090, i believe it will be very helpful for beginners like me.
Can I reuse this project? where i can start from?

**Comment by FiveTechSoft** [comment]
+1

**Comment by nitishpandey04** [comment]
You can do technically. It's just going to take that much more time. 

Here's what @karpathy has to say on this -
https://github.com/karpathy/nanochat?tab=readme-ov-file#bigger-models

**Comment by aospan** [comment]
You can try this pull request - it makes nanochat configurable for various GPU setups, including single 16GB GPUs: https://github.com/karpathy/nanochat/pull/75

**Comment by recursive-duck** [comment]
I'm currently doing it, but the **context length is 1024** instead of 2048:
```
step 25999/171200 (15.19%) | loss: 3.286087 | lrm: 1.00 | dt: 8435.87ms | tok/sec: 242 | mfu: 2.50 | total time: 3659.10m
Step 26000 | Validation bpb: 0.9752
Evaluating: hellaswag_zeroshot (0-shot, type: multiple_choice)... accuracy: 0.3360 | centered: 0.1147 | time: 15.43s
Evaluating: jeopardy (10-shot, type: language_modeling)... accuracy: 0.0080 | centered: 0.0080 | time: 15.99s
Evaluating: bigbench_qa_wikidata (10-shot, type: language_modeling)... accuracy: 0.3900 | centered: 0.3900 | time: 14.32s
Evaluating: arc_easy (10-shot, type: multiple_choice)... accuracy: 0.5420 | centered: 0.3893 | time: 35.94s
Evaluating: arc_challenge (10-shot, type: multiple_choice)... accuracy: 0.2880 | centered: 0.0507 | time: 40.20s
Evaluating: copa (0-shot, type: multiple_choice)... accuracy: 0.5900 | centered: 0.1800 | time: 2.88s
Evaluating: commonsense_qa (10-shot, type: multiple_choice)... accuracy: 0.3100 | centered: 0.1375 | time: 44.42s
Evaluating: piqa (10-shot, type: multiple_choice)... accuracy: 0.6180 | centered: 0.2360 | time: 22.02s
Evaluating: openbook_qa (0-shot, type: multiple_choice)... accuracy: 0.2880 | centered: 0.0507 | time: 14.68s
Evaluating: lambada_openai (0-shot, type: language_modeling)... accuracy: 0.2500 | centered: 0.2500 | time: 14.75s
Evaluating: hellaswag (10-shot, type: multiple_choice)... accuracy: 0.3220 | centered: 0.0960 | time: 85.96s
Evaluating: winograd (0-shot, type: schema)... accuracy: 0.5678 | centered: 0.1355 | time: 7.86s
Evaluating: winogrande (0-shot, type: schema)... accuracy: 0.4980 | centered: -0.0040 | time: 16.99s
Evaluating: bigbench_dyck_languages (10-shot, type: language_modeling)... accuracy: 0.0760 | centered: 0.0760 | time: 22.05s
Evaluating: agi_eval_lsat_ar (3-shot, type: multiple_choice)... accuracy: 0.2478 | centered: 0.0598 | time: 37.39s
Evaluating: bigbench_cs_algorithms (10-shot, type: language_modeling)... accuracy: 0.4160 | centered: 0.4160 | time: 17.72s
Evaluating: bigbench_operators (10-shot, type: language_modeling)... accuracy: 0.0857 | centered: 0.0857 | time: 7.06s
Evaluating: bigbench_repeat_copy_logic (10-shot, type: language_modeling)... accuracy: 0.0000 | centered: 0.0000 | time: 1.09s
Evaluating: squad (10-shot, type: language_modeling)... accuracy: 0.0200 | centered: 0.0200 | time: 59.69s
Evaluating: coqa (0-shot, type: language_modeling)... accuracy: 0.0800 | centered: 0.0800 | time: 19.20s
Evaluating: boolq (10-shot, type: multiple_choice)... accuracy: 0.4300 | centered: -0.5000 | time: 84.92s
Evaluating: bigbench_language_identification (10-shot, type: multiple_choice)... accuracy: 0.2360 | centered: 0.1595 | time: 146.97s
Step 26000 | CORE metric: 0.1105
<|bos|>The capital of France is the city of Paris. It is the capital of the French Republic. The capital
<|bos|>The chemical symbol of gold is the symbol of the chemical element. The symbol of gold is the symbol of the
<|bos|>If yesterday was Friday, then tomorrow will be Thursday, then Wednesday, then Wednesday, then Wednesday, then Wednesday, then Wednesday
<|bos|>The opposite of hot is hot. The opposite of hot is hot. The opposite of hot is hot.
<|bos|>The planets of the solar system are: Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune,
<|bos|>My favorite color is the color of the sky. I like to use the color of the sky to
<|bos|>If 5*x + 3 = 13, then x is the number of the number of digits in the number of digits in the number of
step 26000/171200 (15.19%) | loss: 3.302840 | lrm: 1.00 | dt: 8412.21ms | tok/sec: 243 | mfu: 2.50 | total time: 3659.24m
```
You can estimate how much base_train will take: 24000 minutes or 16-17 days.

**Comment by hdocmsu** [comment]
@AutomaticHourglass, can you elaborate on what steps you took to start training on your computer? Did you use the changes PR (https://github.com/karpathy/nanochat/pull/75)?

**Comment by FiveTechSoft** [comment]
Training on a RTX3060:

https://github.com/karpathy/nanochat/discussions/176

**Comment by recursive-duck** [comment]
@hdocmsu Almost the same, only changed minimal amount of constants in base_train.py to make it happen.

**Comment by nitishpandey04** [comment]
@AutomaticHourglass are you also using 4090 ?

**Comment by sdo2025a** [comment]
Thanks, @AutomaticHourglass!

**Comment by zussini** [comment]
I have run it on rtx4090 according to
#1
So I have run pretraining:
torchrun --standalone --nproc_per_node=1 -m scripts.base_train -- --depth=20
with batch size = 4 and it looks like it will take around 3 days:
step 04245/21400 (19.84%) | loss: 3.012123 | lrm: 1.00 | dt: 13776.73ms | tok/sec: 594 | mfu: 13.44 | total time: 971.74m

Best Regards

**Comment by zussini** [comment]
I am doing it on rtx4090 according to
https://github.com/karpathy/nanochat/discussions/1
So I have run pretraining:
torchrun --standalone --nproc_per_node=1 -m scripts.base_train -- --depth=20
with batch size = 4 and it looks like it will take around 3 days:
step 04245/21400 (19.84%) | loss: 3.012123 | lrm: 1.00 | dt: 13776.73ms | tok/sec: 594 | mfu: 13.44 | total time: 971.74m

Best Regards

**Comment by recursive-duck** [comment]
> [@AutomaticHourglass](https://github.com/AutomaticHourglass) are you also using 4090 ?

No 3090, 4090s are Price for now :(

**Comment by svlandeg** [comment]
Hi all! Love this discussion, it's good to see people getting this to work on low resources. Let me just move this to the discussion forum as it's not really an issue 😎
